compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -2267,26 +2268,66 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                         return None;
                     }
 
-                    // FIXME(fmease): Unsolved vars can escape this InferCtxt snapshot.
-                    Some((assoc_item, def_scope, infcx.resolve_vars_if_possible(impl_substs)))
+                    let impl_substs = if !self.allow_ty_infer() {
+                        let substs = infcx.resolve_vars_if_possible(impl_substs);
+                        assert!(!substs.needs_infer());
+                        Some(substs)
+                    } else {
+                        None
+                    };
+
+                    Some((item, impl_, impl_substs))
                 })
             })
             .collect();
 
         if applicable_candidates.len() > 1 {
             return Err(self.complain_about_ambiguous_inherent_assoc_type(
                 name,
-                applicable_candidates.into_iter().map(|(candidate, ..)| candidate).collect(),
+                applicable_candidates.into_iter().map(|((candidate, ..), ..)| candidate).collect(),
                 span,
             ));
         }
 
-        if let Some((assoc_item, def_scope, impl_substs)) = applicable_candidates.pop() {
+        if let Some(((assoc_item, def_scope), impl_, probe_impl_substs)) =
+            applicable_candidates.pop()
+        {
             self.check_assoc_ty(assoc_item, name, def_scope, block, span);
 
-            // FIXME(inherent_associated_types): To fully *confirm* the *probed* candidate, we still
-            // need to relate the Self-type with fresh item substs & register region obligations for
-            // regionck to prove/disprove.
+            // FIXME(fmease, inherent_associated_types): Register WF obligations for the Self type.
```

Comment:
This would be fixed by representing it as a projection, so we see the `Self` type during wfcheck.

Reply:
I do plan on implementing something akin to `InherentProjection` at least experimentally as we've already discussed [here](https://github.com/rust-lang/rust/pull/105961#discussion_r1067550372). Since I haven't looked into it that much yet, I wasn't sure if this covers wf-checking and inference for the `Self` type or just for the parameters & bounds of an inherent GAT.